### PR TITLE
boards: soc: Set zephyr,flash chosen node for i.MX RT boards

### PR DESCRIPTION
**Pull request by MaureenHelm**
Removes the CODE_LOCATION Kconfig symbol from the i.MX RT SoC series and
refactors the corresponding boards to use a device tree chosen node
zephyr,flash instead. This is similar to commit
0797602 which did the same for
DATA_LOCATION and zephyr,sram.

### CODE REVIEW

File: dts/arm/nxp/nxp_rt6xx_common.dtsi [hunk 1]
```diff
@@ -81,13 +81,13 @@
...
 
-	flexspi: spi@134000 {
+	flexspi: flash-controller@134000 {
 		compatible = "nxp,imx-flexspi";
 		reg = <0x134000 0x1000>;
 		interrupts = <42 0>;
 		label = "FLEXSPI";
 		#address-cells = <1>;
-		#size-cells = <0>;
+		#size-cells = <1>;
```
Comment (mmahadevan108) [inline dts/arm/nxp/nxp_rt6xx_common.dtsi:90]: Could you please help me understand this change.

**Review by jhqian — APPROVED** [review]

**Review by galak — CHANGES_REQUESTED** [review]
The DTS change is not inline with how SPI flash's should be described.  What's the issue in how things are currently?

**Comment by MaureenHelm** [comment]
> The DTS change is not inline with how SPI flash's should be described. What's the issue in how things are currently?

The way we use Kconfig to set the code location isn't inline with the rest of Zephyr. I've been looking at enabling mcuboot on these SoCs, but I need to be able to choose a fixed-partition or the whole flash for the image. That doesn't really work with the Kconfig-ery we have today.

**Review by dleach02 — APPROVED** [review]

**Comment by galak** [comment]
> The way we use Kconfig to set the code location isn't inline with the rest of Zephyr. I've been looking at enabling mcuboot on these SoCs, but I need to be able to choose a fixed-partition or the whole flash for the image. That doesn't really work with the Kconfig-ery we have today.

Why doesn't the partition layout work?  The partition table for the SPI flash should be 0 based, and thus would be an offset from whatever the base address of the qspi region is.  So whatever 'zephyr,code-partition' points to should in theory get added to CONFIG_FLASH_BASE_ADDRESS.

**Comment by arvid-r** [comment]
How do you plan to find the flash device in MCUboot? We run MCUboot on i.MX RT now, but I use `DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL` and a `zephyr,flash-controller = &is25wp064;` in the devicetree. Don't know if that is the right path tough.

**Comment by galak** [comment]
> How do you plan to find the flash device in MCUboot? We run MCUboot on i.MX RT now, but I use `DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL` and a `zephyr,flash-controller = &is25wp064;` in the devicetree. Don't know if that is the right path tough.

Is this necessary to be set?  If so, what does it need to be set to?  Or can this JEDEC_SPI_NOR_0_LABEL?

**Comment by arvid-r** [comment]
I'm not sure, but the relevant code look like this:
```
#if (!defined(CONFIG_XTENSA) && defined(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL))
    if (!flash_device_get_binding(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL)) {
        BOOT_LOG_ERR("Flash device %s not found",
		     DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL);
        while (1)
            ;
    }
#elif (defined(CONFIG_XTENSA) && defined(JEDEC_SPI_NOR_0_LABEL))
    if (!flash_device_get_binding(JEDEC_SPI_NOR_0_LABEL)) {
        BOOT_LOG_ERR("Flash device %s not found", JEDEC_SPI_NOR_0_LABEL);
        while (1)
            ;
    }
#endif
```
Also, @pabigot recently wrote in slack, that:
"First, anything that uses a driver other than the existing spi_nor should not use jedec,spi-nor as a compatible, because that compatible is used for something that's a child of a standard Zephyr SPI bus.  The draft QSPI PRs may do so right now, but they shouldn't."

So, I'm not sure how this is intended to work.

**Comment by pabigot** [comment]
> Also, @pabigot recently wrote in slack, that:
> "First, anything that uses a driver other than the existing spi_nor should not use jedec,spi-nor as a compatible, because that compatible is used for something that's a child of a standard Zephyr SPI bus. The draft QSPI PRs may do so right now, but they shouldn't."

I should clarify that: if the devicetree node for the flash device is structurally a child of a spi bus (in this case, if there's a Zephyr SPI driver for flexspi), *and* the flash device actually can be controlled with the `spi-nor` driver, then `jedec,spi-nor` is appropriate for the compatible.  If the *preferred* driver is something else that uses QSPI, say `nxp,flexspi-nor`, and again the contents of the node are consistent with that binding, then we'd want something like:

    		compatible = "adesto,at25sf128a", "nxp,flexspi", "jedec,spi-nor";
￼
allowing somebody to have, in preferred order:
* a driver specific to the adesto hardware; or
* the flexspi QSPI-aware driver; or
* the standard SPI driver

depending on which drivers were enabled by Kconfig.

**Review by pabigot — APPROVED** [review]
AFAICT this is just moving the canonical source for existing flash-related data from Kconfig to devicetree, which seems reasonable.

Looking at this `nxp,imx-flexspi` is being described as a flash controller, so in that case `jedec,spi-nor` should be removed as soon as support for using the flash devices in Zephyr is complete.

**Comment by arvid-r** [comment]
Thanks!

**Review by mbolivar-nordic — APPROVED** [review]

File: dts/arm/nxp/nxp_rt6xx_common.dtsi (hunk 1 above)
Comment (nvlsianpu) [reply dts/arm/nxp/nxp_rt6xx_common.dtsi:90]: #size-cell need to be set to 1 in order to tell DTC that register size should be emitted as a 32-bit value.

**Review by nvlsianpu — APPROVED** [review]

**Review by eHammarstrom — APPROVED** [review]

**Comment by MaureenHelm** [comment]
Superseded by #34845

**Pull request closed without merging**